Support to get github api from specified json path and custom asset host prefix

newbe36524 opened the pull request:
Support to get github api from specified json path and custom asset host prefix

# Description

1. add $DaprReleaseJson  to specfied github api responce json
2. add $CustomAssetPrefix to replace github.com with specified custom asset host
3. download github release file with browser_download_url instead of url with api

for example, developer could run this script to download and install dapr in China to get better network expreince:

```powershell
./install.ps1 -DaprReleaseJson "https://gitee.com/dapr-cn/dapr-bin-mirror/raw/main/cli/releases.json" -CustomAssetPrefix "https://ghproxy.com/https://github.com/dapr/cli/releases/download/"
```


## Issue reference

close: #647

## Checklist

* [X] Code compiles correctly
* [X] Created/updated tests
* [X] Extended the documentation

mukundansundar commented:
@yaron2 Can you take a look at this PR?

artursouza commented on install/install.ps1:
```diff
@@ -69,18 +76,22 @@ if (!$Version) {
 Write-Output "Downloading $zipFileUrl ..."
 
 $githubHeader.Accept = "application/octet-stream"
-Invoke-WebRequest -Headers $githubHeader -Uri $zipFileUrl -OutFile $zipFilePath
```
What if URL is from GitHub, don't this fail due to missing header? Or, at least have higher risk of being throttled.

newbe36524 replied:
https://github.com/dapr/cli/blob/master/install/install.sh#L109

In fact, I thought this would work because I saw that in the install.sh version, no specific header was passed when downloading the file.
Also, I have tested locally that it works without passing this header to download resources from github.
If you think there is a better way to verify this, feel free to ask.

newbe36524 replied:
And the browser_download_url version is the one used in install.sh, which is the domain name starting with github.com, not the one starting with api.github.com. So I think it should work.

artursouza replied:
It might work for some time. On the other hand, throttling will happen pretty quickly without GitHub Auth token. Any reason for this to be removed?

newbe36524 replied:
I think it works in the same way as install.sh that request to github.com directly instead of api.github.com.
If there is throttling, it should happen to current version of install.sh now published.

artursouza replied:
We can fix `install.sh` in another PR. The fact that this problem exists in `install.sh` does not mean it should exist here. We are seeing issues with github API throttles in other parts of the code and I don't see a reason to add more to the problem.

Please, put the header back.

yaron2 commented:
ping @newbe36524

newbe36524 commented:
@artursouza Thanks to your review and there are some updated. Please help me to review it.

- I fixed the header issue as you mentioned
- I've updated Readme.md to include instructions on how to customize use this script

artursouza commented:
Thanks, @newbe36524. I will be validating this PR this week before merging. This script is critical, so I will confirm the GitHub path is not broken.

berndverst commented on install/README.md:
```diff
@@ -41,3 +41,62 @@ wget -q https://raw.githubusercontent.com/dapr/cli/master/install/install.sh -O
+    CustomAssetFactory = ${function:CustomAssetFactory};
+    DaprReleaseJsonUrl    = "https://gitee.com/dapr-cn/dapr-bin-mirror/raw/main/cli/releases.json";
+}
+$script=iwr -useb https://cdn.jsdelivr.net/gh/dapr/cli/installer/install.ps1;
```
Replace URL with `https://cdn.jsdelivr.net/gh/dapr/cli/install/install.ps1` as this URL doesn't work

```suggestion
$script=iwr -useb https://cdn.jsdelivr.net/gh/dapr/cli/install/install.ps1;
```

berndverst commented on install/README.md:
```diff
@@ -41,3 +41,62 @@ wget -q https://raw.githubusercontent.com/dapr/cli/master/install/install.sh -O
+    DaprReleaseJsonUrl    = "https://gitee.com/dapr-cn/dapr-bin-mirror/raw/main/cli/releases.json";
+    Version            = <Version>
+}
+$script=iwr -useb https://cdn.jsdelivr.net/gh/dapr/cli/installer/install.ps1;
```
Replace URL with `https://cdn.jsdelivr.net/gh/dapr/cli/install/install.ps1` as this URL doesn't work

```suggestion
$script=iwr -useb https://cdn.jsdelivr.net/gh/dapr/cli/install/install.ps1;
```

berndverst commented on install/README.md:
```diff
@@ -41,3 +41,62 @@ wget -q https://raw.githubusercontent.com/dapr/cli/master/install/install.sh -O
+### Windows
+
+- Create a `CustomAssetFactory` function to define what the release asset url you want to use
+- Pass `DaprReleaseJsonUrl` as the json result of all releases as <https://api.github.com/repos/dapr/cli/releases>
```
Wording is confusing. Suggestion::

```suggestion
- Set `DaprReleaseJsonUrl` to the equivalent of the json representation of all releases at <https://api.github.com/repos/dapr/cli/releases>
```

berndverst commented on install/install.ps1:
```diff
@@ -58,23 +60,61 @@ if (!(Test-Path $DaprRoot -PathType Container)) {
-    $windowsAsset = $releases | Where-Object { $_.tag_name -notlike "*rc*" } | Select-Object -First 1 | Select-Object -ExpandProperty assets | Where-Object { $_.name -Like "*windows_amd64.zip" }
-    if (!$windowsAsset) {
-        throw "Cannot find the windows Dapr CLI binary"
+# get latest or spcified version info from releases
```
typo should be `specified`

```suggestion
# get latest or specified version info from releases
```

berndverst reviewed (dismissed):
LGTM

Just make sure to fix a few typos and the README (including the example).

berndverst commented:
FYI @artursouza this looks good.

newbe36524 commented:
> LGTM
> 
> Just make sure to fix a few typos and the README (including the example).

Thanks and updated
